### ISSUE TITLE
Softmax Sampling Support

### ISSUE THREAD

**Issue by wchan** [tensorflow/tensorflow#565]
I couldn't find all the pieces to do scheduled softmax sampling (Bengio et al, 2015), is that supported in TF atm? Is there an TensorOp that can sample from the previous layer?

Thanks!

**Comment by ludimagister** [comment]
Take a look at this section:

https://www.tensorflow.org/versions/master/api_docs/python/nn.html#candidate-samplers

The sampled softmax loss is here:

https://github.com/tensorflow/tensorflow/blob/20723e2b3d58cc48b2a302f7ea9806c8a75fd18f/tensorflow/python/ops/nn.py#L778

For example used here:

https://github.com/tensorflow/tensorflow/blob/9c3043ff3bf31a6a81810b4ce9e87ef936f1f529/tensorflow/models/rnn/translate/seq2seq_model.py#L98

**Comment by wchan** [comment]
I think that's a different "sampled softmax" ;) , the one I was referring to is to condition on a sample (either from the model, or from the ground truth; see Bengio et al, NIPS 2015), as opposed to the "sampled softmax" for estimating the partition function.

**Comment by wchan** [comment]
FYI, for those looking for a quick op for this, i wrote one here:
master/tensorflow/core/kernels/gru_op.cc

its under "TokenSampleOp", lots of code refactoring and generalization and tests needed before a patch can be submitted "soon"

**Comment by girving** [comment]
@ludimagister, @wchan: From the discussion above it seems like this was a resolved question, especially since #2093 is closed.  Please let me know if we should reopen!

**Comment by JasonZhao001** [comment]
it will be helpful:
https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/Categorical
